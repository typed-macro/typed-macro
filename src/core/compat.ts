/*
 * This is a compatibility control mechanism based on version number,
 * to ensure the compatibility between macro and runtime.
 */

import { Macro } from '@/core/macro'
import { WithVersion } from '@/common'

/*
 * The current Macro version.
 *
 * It represents whether the call-context passed by the Runtime to the Macro is compatible.
 * This involves not only the structure of the call-context itself,
 * but also the structure of the AST between different versions of Babel,
 * for AST is generated in Runtime but transformed in Macros.
 *
 * @deps {@link RawMacroCall}, {@link Babel}
 */
export const MACRO_VERSION = 0

export function versionedMacro(v: Macro): Macro {
  ;(v as WithVersion<Macro>).$__version = MACRO_VERSION
  return v
}

export function isMacroCompatible(v: Macro) {
  return (v as WithVersion<Macro>).$__version === MACRO_VERSION
}
