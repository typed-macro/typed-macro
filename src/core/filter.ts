import { posix, win32 } from 'path'
import pm from 'picomatch'

export type FilterPattern = RegExp | string | (RegExp | string)[]

export function createFilter(include?: FilterPattern, exclude?: FilterPattern) {
  const getMatcher = (pattern: string | RegExp) => {
    if (pattern instanceof RegExp) return pattern
    const fn = pm(pattern, { dot: true })
    return { test: (id: string) => fn(id) }
  }
  const ensureArray = (v: unknown) => (v ? (Array.isArray(v) ? v : [v]) : [])

  const includeMatchers = ensureArray(include).map(getMatcher)
  const excludeMatchers = ensureArray(exclude).map(getMatcher)

  return (id: string) => {
    const pathId = id.split(win32.sep).join(posix.sep)
    for (let i = 0; i < excludeMatchers.length; ++i) {
      if (excludeMatchers[i].test(pathId)) return false
    }

    for (let i = 0; i < includeMatchers.length; ++i) {
      if (includeMatchers[i].test(pathId)) return true
    }

    return !includeMatchers.length
  }
}
