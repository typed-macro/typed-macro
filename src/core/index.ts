export { Runtime, RuntimeOptions } from './runtime'
export {
  createTransformer,
  Transformer,
  TransformerOptions,
} from './transformer'
export {
  Macro,
  MacroContext,
  MacroHelper,
  MacroHandler,
  Exportable,
  NamespacedExportable,
  NormalizedExports,
  normalizeExports,
} from './types'
