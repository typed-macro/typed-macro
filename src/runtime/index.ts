import {
  NamespacedMacros,
  NamespacedModules,
  NamespacedTypes,
  NormalizedExports,
  renderTypes,
} from '@/runtime/types'
import {
  createTransformer,
  Transformer,
  TransformerOptions,
} from '@/runtime/transformer'
import { findDuplicatedItem } from '@/common'
import { mkdir, writeFile } from 'fs/promises'
import { dirname } from 'path'

export type RuntimeOptions = {
  transformer: TransformerOptions
}

export class Runtime {
  private macros: NamespacedMacros = Object.create(null)
  private modules: NamespacedModules = Object.create(null)
  private types: NamespacedTypes = Object.create(null)

  private macrosNamespaces: string[] = []
  private modulesNamespaces: string[] = []

  private devMode = false

  private readonly transformer: Transformer

  constructor(options: RuntimeOptions) {
    this.transformer = createTransformer(options.transformer)
  }

  setDevMode(dev = true) {
    this.devMode = dev
  }

  register({ macros, modules, types }: NormalizedExports) {
    {
      const duplicated = findDuplicatedItem(
        Object.keys(types),
        Object.keys(this.types)
      )
      if (duplicated) throw new Error(`duplicated namespace '${duplicated}'.`)
    }
    {
      Object.keys(macros).forEach((ns) => {
        const mem = Object.create(null)
        macros[ns].forEach((m) => {
          if (mem[m.name]) {
            throw new Error(
              `a macro with name '${m.name}' in '${ns}' already existed`
            )
          }
          mem[m.name] = 1
        })
      })
    }
    Object.assign(this.macros, macros)
    this.macrosNamespaces = Object.keys(this.macros)
    Object.assign(this.modules, modules)
    this.modulesNamespaces = Object.keys(this.modules)
    Object.assign(this.types, types)
  }

  handleLoad(id: string) {
    if (this.macrosNamespaces.includes(id)) return 'export {}'
    if (this.modulesNamespaces.includes(id)) return this.modules[id]
  }

  handleResolveId(id: string) {
    if (
      this.macrosNamespaces.includes(id) ||
      this.modulesNamespaces.includes(id)
    )
      return id
  }

  handleTransform(code: string, id: string, ssr = false) {
    if (/\.[jt]sx?$/.test(id))
      return this.transformer(
        {
          code,
          id,
          ssr,
          dev: this.devMode,
        },
        this.macros
      )
  }

  get container() {
    return {
      macros: this.macros,
      modules: this.modules,
      types: this.types,
    }
  }

  async generateDts(path: string) {
    await mkdir(dirname(path), { recursive: true })
    await writeFile(path, renderTypes(this.types))
  }
}
