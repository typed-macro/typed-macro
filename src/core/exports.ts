import { isMacro, Macro } from '@/core/macro'
import { isMacroCompatible } from '@/core/compat'

export type NamespacedMacros = { [namespace: string]: Macro[] }

export type NamespacedModules = { [namespace: string]: string }

export type NamespacedTypes = {
  [namespace: string]: {
    moduleScope: string[]
    macroScope: string[]
  }
}

export type Exportable = (
  | {
      /**
       * as macros
       */
      macros: Macro[]
    }
  | {
      /**
       * as module, code will be used as source code of the module
       */
      code: string
    }
) & {
  /**
   * Type definitions, will be written to d.ts.
   *
   * e.g.
   * ```typescript
   * { exports: { '@macros': { customTypes: `type A = string` } } }
   * ```
   * will generate
   * ```typescript
   * declare module '@macros' {
   *   type A = string
   * }
   * ```
   */
  customTypes?: string
}

export type NamespacedExportable = { [namespace: string]: Exportable }

export type NormalizedExports = {
  macros: NamespacedMacros
  modules: NamespacedModules
  types: NamespacedTypes
}

export function normalizeExports(
  exports: NamespacedExportable
): NormalizedExports {
  const macros: NamespacedMacros = Object.create(null)
  const modules: NamespacedModules = Object.create(null)
  const types: NamespacedTypes = Object.create(null)
  Object.keys(exports).forEach((ns) => {
    const item = exports[ns]
    types[ns] = {
      moduleScope: item.customTypes ? [item.customTypes] : [],
      macroScope: [],
    }
    if ('code' in item) {
      modules[ns] = item.code
    } else {
      macros[ns] = item.macros
      types[ns].macroScope = item.macros.map((m) => {
        if (!isMacro(m)) throw new Error(`'${m.name}' is not a macro`)
        return m.__types
      })
    }
  })

  validateMacros(macros)

  return {
    macros,
    modules,
    types,
  }
}

// ensure compatible and no name conflict
export function validateMacros(macros: NamespacedMacros) {
  Object.keys(macros).forEach((ns) => {
    const mem = Object.create(null)
    macros[ns].forEach((m) => {
      if (!isMacroCompatible(m))
        throw new Error(`macro '${m.name}' in '${ns}' is incompatible`)

      if (mem[m.name])
        throw new Error(`macro '${m.name}' in '${ns}' is duplicated`)

      mem[m.name] = 1
    })
  })
}
