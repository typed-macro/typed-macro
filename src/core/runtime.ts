import {
  createTypeRenderer,
  TypeRenderer,
  TypeRendererOptions,
} from '@/core/typeRenderer'
import {
  createTransformer,
  Transformer,
  TransformerOptions,
} from '@/core/transformer'
import {
  NamespacedMacros,
  NamespacedModules,
  NamespacedTypes,
  NormalizedExports,
  validateMacros,
} from './exports'
import { DeepPartial, findDuplicatedItem } from '@/common'
import { createFilter, FilterOptions } from './filter'

export type RuntimeOptions = {
  /**
   * The options for transformer.
   */
  transformer: TransformerOptions
  /**
   * The options for typeRenderer.
   */
  typeRenderer: Pick<TypeRendererOptions, 'typesPath'>
  /**
   * The options that control the filtering of which files
   * require transformations to be applied.
   */
  filter: FilterOptions
}

export type Attachable = {
  exports: NormalizedExports
  options?: DeepPartial<RuntimeOptions>
}

export class Runtime {
  private macros: NamespacedMacros = Object.create(null)
  private modules: NamespacedModules = Object.create(null)
  private types: NamespacedTypes = Object.create(null)

  private macrosNamespaces: string[] = []
  private modulesNamespaces: string[] = []

  private devMode = false

  constructor(
    private readonly _options: RuntimeOptions,
    defaultExports?: NormalizedExports
  ) {
    if (defaultExports) this.addExports(defaultExports)
  }

  setDevMode(dev = true) {
    this.devMode = dev
  }

  addExports({ macros, modules, types }: NormalizedExports) {
    assertNoDuplicatedNamespace(Object.keys(this.types), Object.keys(types))
    assertNoDuplicatedNamespace(this.macrosNamespaces, Object.keys(macros))
    assertNoDuplicatedNamespace(this.modulesNamespaces, Object.keys(modules))
    validateMacros(macros)
    Object.assign(this.macros, macros)
    this.macrosNamespaces = Object.keys(this.macros)
    Object.assign(this.modules, modules)
    this.modulesNamespaces = Object.keys(this.modules)
    Object.assign(this.types, types)
  }

  handleLoad(id: string) {
    if (this.macrosNamespaces.includes(id)) return 'export {}'
    if (this.modulesNamespaces.includes(id)) return this.modules[id]
  }

  handleResolveId(id: string) {
    if (
      this.macrosNamespaces.includes(id) ||
      this.modulesNamespaces.includes(id)
    )
      return id
  }

  private _transformer?: Transformer
  private _typeRenderer?: TypeRenderer
  private _filter?: (id: string) => boolean

  get attachable() {
    return {
      options: this._options,
      exports: {
        macros: this.macros,
        modules: this.modules,
        types: this.types,
      },
    }
  }

  get filter() {
    return this._filter || (this._filter = createFilter(this._options.filter))
  }

  get transformer() {
    return (
      this._transformer ||
      (this._transformer = createTransformer(this._options.transformer))
    )
  }

  get typeRenderer() {
    return (
      this._typeRenderer ||
      (this._typeRenderer = createTypeRenderer({
        types: this.types,
        typesPath: this._options.typeRenderer.typesPath,
      }))
    )
  }

  handleTransform(code: string, filepath: string, ssr = false) {
    if (this.filter(filepath))
      return this.transformer(
        {
          code,
          filepath,
          ssr,
          dev: this.devMode,
        },
        this.macros
      )
  }

  mergeOptions(options: DeepPartial<RuntimeOptions>) {
    if (options.transformer) {
      // reset transformer
      this._transformer = undefined
      // merge transformer#parserPlugins
      if (options.transformer.parserPlugins?.length) {
        this._options.transformer.parserPlugins = Array.from(
          new Set([
            ...(this._options.transformer.parserPlugins || []),
            ...options.transformer.parserPlugins,
          ])
        )
      }
    }
  }

  attach({ exports, options }: Attachable) {
    // merge exports
    this.addExports(exports)
    // merge options
    options && this.mergeOptions(options)
  }
}

export function assertNoDuplicatedNamespace(ns1: string[], ns2: string[]) {
  const duplicated = findDuplicatedItem(ns1, ns2)
  if (duplicated) throw new Error(`duplicated namespace '${duplicated}'.`)
}
