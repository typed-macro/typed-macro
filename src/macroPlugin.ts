import type { Plugin, ViteDevServer } from 'vite'
import { NormalizedExports } from '@/core/exports'
import { MacroProvider } from '@/macroProvider'
import { DevServerHelper, getDevServerHelper } from '@/helper/server'
import { Runtime, RuntimeOptions } from '@/core/runtime'

export type MacroPluginHooks = Omit<
  Plugin,
  'name' | 'enforce' | 'apply' | 'configureServer'
> & {
  configureServer?: (
    server: ViteDevServer,
    helpers: DevServerHelper
  ) => (() => void) | void | Promise<(() => void) | void>
}

export type InternalPluginOptions = {
  name: string
  runtimeOptions: RuntimeOptions
  exports: NormalizedExports
  hooks: MacroPluginHooks
}

export interface MacroPlugin extends Plugin {
  __consume: () => MacroProvider
}

interface InternalMacroPlugin extends MacroPlugin {
  __internal_macro_plugin: true
}

export function macroPlugin(options: InternalPluginOptions): MacroPlugin {
  const {
    name,
    exports,
    runtimeOptions,
    hooks: {
      buildStart,
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    },
  } = options

  let runtime: Runtime | undefined = new Runtime(runtimeOptions)

  runtime.register(exports)

  return {
    __internal_macro_plugin: true,
    __consume() {
      if (!runtime) throw new Error(`plugin '${name}' is used more than once.`)
      const provider: MacroProvider = {
        id: name,
        exports: runtime.exports,
        hooks: {},
      }
      runtime = undefined
      return provider
    },
    name,
    enforce: 'pre',
    async buildStart(opt) {
      await Promise.all([
        runtime?.typeRenderer.write(),
        buildStart?.bind(this)(opt),
      ])
    },
    configResolved(config) {
      runtime?.setDevMode(config?.env?.DEV ?? false)
      return configResolved?.(config)
    },
    resolveId(id, importer, options, ssr) {
      const result = runtime?.handleResolveId(id)
      if (result) return result
      return resolveId?.bind(this)(id, importer, options, ssr)
    },
    load(id, ssr) {
      const result = runtime?.handleLoad(id)
      if (result) return result
      return load?.bind(this)(id, ssr)
    },
    transform(code, id, ssr) {
      const transformed = runtime?.handleTransform(code, id, ssr)
      return transform?.bind(this)(transformed ?? code, id, ssr) ?? transformed
    },
    configureServer(server) {
      // hook
      if (configureServer) configureServer(server, getDevServerHelper(server))
    },
    ...otherHooks,
  } as InternalMacroPlugin
}

export function isMacroPlugin(o: unknown): o is MacroPlugin {
  return (o as any).__internal_macro_plugin ?? false
}
