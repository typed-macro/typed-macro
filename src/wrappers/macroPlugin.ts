import type { Plugin, ViteDevServer } from 'vite'
import { DevServerHelper, getDevServerHelper } from '@/wrappers/helper/server'
import { Attachable, Runtime } from '@/core/runtime'
import { versionedPlugin } from '@/wrappers/compat'

export type MacroPluginHooks = Omit<
  Plugin,
  'name' | 'enforce' | 'apply' | 'configureServer'
> & {
  configureServer?: (
    server: ViteDevServer,
    helpers: DevServerHelper
  ) => (() => void) | void | Promise<(() => void) | void>
}

export type InternalPluginOptions = {
  name: string
  runtime: Runtime
  hooks: MacroPluginHooks
}

export interface MacroPlugin extends Plugin {
  __consume: () => Attachable
}

interface InternalMacroPlugin extends MacroPlugin {
  __internal_macro_plugin: true
}

export function macroPlugin(options: InternalPluginOptions): MacroPlugin {
  const {
    name,
    hooks: {
      buildStart,
      configResolved,
      configureServer,
      transform,
      load,
      resolveId,
      ...otherHooks
    },
  } = options

  let runtime: Runtime | undefined = options.runtime

  const plugin = {
    __internal_macro_plugin: true,
    __consume() {
      if (!runtime) throw new Error(`plugin '${name}' is used more than once.`)
      const consume = {
        exports: runtime.exports,
        options: runtime.options,
      }
      runtime = undefined
      return consume
    },
    name,
    enforce: 'pre',
    configResolved(config) {
      runtime?.setDevMode(config?.env?.DEV ?? false)
      return configResolved?.(config)
    },
    configureServer(server) {
      if (configureServer) configureServer(server, getDevServerHelper(server))
    },
    async buildStart(opt) {
      await Promise.all([
        runtime?.typeRenderer.write(),
        buildStart?.call(this, opt),
      ])
    },
    resolveId(id, importer, options, ssr) {
      const result = runtime?.handleResolveId(id)
      if (result) return result
      return resolveId?.call(this, id, importer, options, ssr)
    },
    load(id, ssr) {
      const result = runtime?.handleLoad(id)
      if (result) return result
      return load?.call(this, id, ssr)
    },
    transform(code, id, ssr) {
      const transformed = runtime?.handleTransform(code, id, ssr)
      return (
        transform?.call(this, transformed ?? code, id, ssr) ??
        (transformed && { code: transformed, map: null })
      )
    },
    ...otherHooks,
  } as InternalMacroPlugin
  return versionedPlugin(plugin)
}

export function isMacroPlugin(o: unknown): o is MacroPlugin {
  return (o as any)?.__internal_macro_plugin ?? false
}
