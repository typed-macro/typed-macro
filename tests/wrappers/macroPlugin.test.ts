import { Runtime } from '@/core/runtime'
import { isMacroPlugin, MacroPlugin, macroPlugin } from '@/wrappers/macroPlugin'
import {
  macroSerializer,
  mockMacro,
  mockRuntime,
  withDevServer,
  withTempPath,
} from '../testutils'

expect.addSnapshotSerializer(macroSerializer)

describe('macroPlugin() & isMacroPlugin()', () => {
  it('should work', () => {
    expect(
      isMacroPlugin(
        macroPlugin({
          hooks: {},
          name: 'test',
          runtime: new Runtime({
            typeRenderer: { typesPath: '' },
            transformer: {},
            filter: {},
          }),
        })
      )
    ).toBe(true)
    expect(isMacroPlugin(undefined)).toBe(false)
    expect(isMacroPlugin(false)).toBe(false)
  })
})

describe('MacroPlugin', () => {
  it('should support to be consumed', () => {
    const plugin = macroPlugin({
      name: 'test',
      hooks: {},
      runtime: mockRuntime(
        {},
        {
          modules: {
            '@helper': `export function hello(msg) {console.log(msg)}`,
          },
          macros: {
            '@echo': [
              mockMacro('echo', ({ path }, { template }, { prependToBody }) => {
                prependToBody(
                  template.statements.ast(
                    `import { hello } from '@helper'; hello('world')`
                  )
                )
                path.remove()
              }),
            ],
          },
          types: {},
        }
      ),
    })
    expect(plugin.__consume()).toMatchSnapshot()
    // consume more than once
    expect(() => plugin.__consume()).toThrow()
  })
})

describe('MacroPlugin Hooks', () => {
  let mockedHook: jest.Mock
  let plugin: MacroPlugin
  beforeEach(() => {
    mockedHook = jest.fn()
    plugin = macroPlugin({
      name: 'test',
      hooks: {
        transform: mockedHook,
        resolveId: mockedHook,
        load: mockedHook,
        configureServer: mockedHook,
        configResolved: mockedHook,
        buildStart: mockedHook,
      },
      runtime: mockRuntime(
        {},
        {
          modules: {
            '@helper': `export function hello(msg) {console.log(msg)}`,
          },
          macros: {
            '@echo': [
              mockMacro('echo', ({ path }, { template }, { prependToBody }) => {
                prependToBody(
                  template.statements.ast(
                    `import { hello } from '@helper'; hello('world')`
                  )
                )
                path.remove()
              }),
            ],
          },
          types: {},
        }
      ),
    })
  })

  it('should handle transform with macro', () => {
    expect(
      plugin.transform!.call(null as any, `import { echo } from '@echo'`, '')
    ).toBeUndefined()
    expect(
      plugin.transform!.call(
        null as any,
        `import { echo } from '@echo'; echo()`,
        'a.ts'
      )
    ).toMatchSnapshot()
    expect(mockedHook.mock.calls.length).toBe(2)
  })

  it('should handle load properly', () => {
    expect(plugin.load!.call(null as any, '@helper')).toMatchSnapshot()
    expect(plugin.load!.call(null as any, '@echo')).toMatchSnapshot()

    expect(plugin.load!.call(null as any, '@others')).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(1)
  })

  it('should handle resolveId properly', () => {
    expect(plugin.resolveId!.call(null as any, '@helper', '', {})).toBe(
      '@helper'
    )
    expect(plugin.resolveId!.call(null as any, '@echo', '', {})).toBe('@echo')

    expect(
      plugin.resolveId!.call(null as any, '@others', '', {})
    ).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(1)
  })

  it('should handle buildStart properly', async () => {
    await withTempPath('a.d.ts', async (tempPath) => {
      plugin = macroPlugin({
        name: 'test',
        runtime: mockRuntime({ typeRenderer: { typesPath: tempPath } }),
        hooks: { buildStart: mockedHook },
      })
      await plugin.buildStart!.call(null as any, {} as any)
      expect(mockedHook.mock.calls.length).toBe(1)
    })
  })

  it('should handle configResolved properly', async () => {
    await plugin.configResolved!.call(null as any, {} as any)
    expect(mockedHook.mock.calls.length).toBe(1)
  })

  it('should handle configureServer properly', async () => {
    await withDevServer(async (server) => {
      await plugin.configureServer!.call(null as any, server)
      expect(mockedHook.mock.calls.length).toBe(1)
    })
  })

  // __CONSUMED__

  it('should stop handling transform after consumed', () => {
    plugin.__consume()
    expect(
      plugin.transform!.call(
        null as any,
        `import { echo } from '@echo'; echo()`,
        'a.ts'
      )
    ).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(1)
  })

  it('should stop handling load after consumed', () => {
    plugin.__consume()
    expect(plugin.load!.call(null as any, '@helper')).toBeUndefined()
    expect(plugin.load!.call(null as any, '@echo')).toBeUndefined()
    expect(plugin.load!.call(null as any, '@others')).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(3)
  })

  it('should stop handling resolveId after consumed', () => {
    plugin.__consume()
    expect(
      plugin.resolveId!.call(null as any, '@helper', '', {})
    ).toBeUndefined()
    expect(plugin.resolveId!.call(null as any, '@echo', '', {})).toBeUndefined()
    expect(
      plugin.resolveId!.call(null as any, '@others', '', {})
    ).toBeUndefined()
    expect(mockedHook.mock.calls.length).toBe(3)
  })
})
