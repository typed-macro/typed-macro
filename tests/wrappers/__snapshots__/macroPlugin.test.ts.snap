// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MacroPlugin Hooks should handle load properly 1`] = `"export function hello(msg) {console.log(msg)}"`;

exports[`MacroPlugin Hooks should handle load properly 2`] = `"export {}"`;

exports[`MacroPlugin Hooks should handle transform with macro 1`] = `"import { hello } from '@helper';hello('world');"`;

exports[`MacroPlugin should support to be consumed 1`] = `
Object {
  "exports": Object {
    "macros": Object {
      "@echo": Array [
        Macro { "name": "echo", "__types": "" },
      ],
    },
    "modules": Object {
      "@helper": "export function hello(msg) {console.log(msg)}",
    },
    "types": Object {},
  },
  "options": Object {
    "filter": Object {},
    "transformer": Object {},
    "typeRenderer": Object {
      "typesPath": "",
    },
  },
}
`;
